src/execution_plan/execution_plan.c

```diff
@@ -108,30 +108,29 @@ Vector* _ExecutionPlan_Locate_References(OpBase *root, OpBase **op, Vector *refe
     /* See if all references been resolved.
      * TODO: create a vector compare function
      * which checks if the content of one is in the other. */
-    int match = Vector_Size(references);
-    size_t ref_count = Vector_Size(references);
-    size_t seen_count = Vector_Size(seen);
-
-    for(int i = 0; i < ref_count; i++) {
-        char *reference;        
-        Vector_Get(references, i, &reference);
+    int match = raxSize(references);
+    int ref_count = match;
+    int seen_count = Vector_Size(seen);
+    
+    // We've seen enough to start matching.
+    if(seen_count >= ref_count) {
+        for(int i = 0; i < seen_count; i++) {
+            // To many unmatched references.
+            if(match > (seen_count-i)) break;
 
-        int j = 0;
-        for(; j < seen_count; j++) {
             char *resolved;
-            Vector_Get(seen, j, &resolved);
-            if(!strcmp(reference, resolved)) {
-                /* Match! */
-                break;
+            Vector_Get(seen, i, &resolved);
+            if(raxFind(references, (unsigned char*)resolved, strlen(resolved)) != raxNotFound) {
```

Comment:
This can cause invalid accesses in queries that have WITH clauses and later WHERE clauses, like:
`MATCH (p:person) WITH avg(p.age) AS average_age MATCH(q:person) WHERE q.age > average_age RETURN q`
The alias `p` is freed after the first AST is converted, but a pointer to it is still stored in the `modifies` Vector, and will be checked when converting the second AST.

The same issue pops up when applying some optimizations.

Reply:
Can you please walk me through where exactly does the leak occurs,
an alias within a `modifies` vector is a pointer into an alias on an entity within a querygraph, in the case of `WITH` the query graph of a sub-query isn't free. See _ExecutionPlan_Connect at the very end of the function, which present a leak by itself.

Reply:
Sure thing!

This isn't a leak, it's an invalid read of a string that's already been freed. Let's take slightly simpler queries like:
`"MATCH (will_be_freed) WITH ID(will_be_freed) AS id MATCH (q) WHERE ID(q) = id RETURN q"`
`"MATCH (will_be_freed) WITH ID(will_be_freed) AS id RETURN id"`

1) We process the first AST, and its QueryGraph consists of one node with the alias "will_be_freed".
2) Then we build an AllNodeScan, and push that alias (the actual pointer, not a `strdup`) onto its `modifies` Vector.
3) We're done with the first AST, and move on to another AST. Early on we call `_ExecutionPlan_Connect`, which as you point out has a memory leak in which we try to avoid double-freeing QueryGraphs.
4) However, the ExecutionPlan now has a bunch of QueryGraphs in `plan->connected_components`, and we still free all of those properly, including the Node and its alias "will_be_freed". 
5) Any access to the `op->modifies` strings on operations built from earlier query graphs are now invalid. This can be a `GRAPH.EXPLAIN`, the `ExecutionPlan_Locate_References` associated with a FilterTree, and I think some optimization applications.

Here are two explain outputs showing invalid string accesses:

```
"MATCH (will_be_freed) WITH ID(will_be_freed) AS id MATCH (q) WHERE ID(q) = id RETURN q"
1) "Results"
2) "    Project"
3) "        Filter"
4) "            Cartesian Product"
5) "                All Node Scan | (q)"
6) "                Project"
7) "                    All Node Scan | (`_\x90UUU)"
```
```
"MATCH (will_be_freed) WITH ID(will_be_freed) AS id RETURN id"
1) "Results"
2) "    Project"
3) "        Project"
4) "            All Node Scan | ()"
```
Trace:
```
==29755==  Address 0x602fe88 is 8 bytes inside a block of size 56 free'd
==29755==    at 0x4C30D3B: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==29755==    by 0x849A408: Node_Free (node.c:126)
==29755==    by 0x84989A6: QueryGraph_Free (query_graph.c:499)
==29755==    by 0x8457B04: ExecutionPlanFree (execution_plan.c:753)
==29755==    by 0x845755C: _ExecutionPlan_Connect (execution_plan.c:647)
==29755==    by 0x8457616: NewExecutionPlan (execution_plan.c:658)
==29755==    by 0x844DBEC: MGraph_Explain (cmd_explain.c:93)
```
Fixing just the access to the alias is easy, but I wonder if there are more dangerous accesses to Node or Edge objects that have been freed in earlier QueryGraphs.

Reply:
👍 thanks!